google/cloud/pubsub/message_test.cc

```diff
@@ -0,0 +1,177 @@
+#include <google/protobuf/text_format.h>
+#include <gmock/gmock.h>
+#include <sstream>
+#include <tuple>
+
+namespace google {
+namespace cloud {
+namespace pubsub {
+inline namespace GOOGLE_CLOUD_CPP_PUBSUB_NS {
+namespace {
+
+using ::google::cloud::testing_util::IsProtoEqual;
+using ::testing::HasSubstr;
+using ::testing::UnorderedElementsAre;
+
+TEST(Message, FromData) {
+  auto const m0 = Message::FromData("contents-0");
+  EXPECT_EQ("contents-0", m0.data());
+  EXPECT_TRUE(m0.attributes().empty());
+  EXPECT_TRUE(m0.ordering_key().empty());
+  EXPECT_TRUE(m0.message_id().empty());
+
+  auto const m1 = Message::FromData("contents-1");
+  EXPECT_EQ("contents-1", m1.data());
+
+  EXPECT_NE(m0, m1);
+  Message copy(m0);
+  EXPECT_EQ(m0, copy);
+  Message move(std::move(copy));
+  EXPECT_EQ(m0, move);
+}
+
+TEST(Message, FromAttributesIterator) {
+  std::map<std::string, std::string> const attributes(
+      {{"k1", "v1"}, {"k2", "v2"}});
+
+  auto const m0 = Message::FromAttributes(attributes.begin(), attributes.end());
+  EXPECT_TRUE(m0.data().empty());
+  EXPECT_THAT(m0.attributes(),
+              UnorderedElementsAre(std::make_pair("k1", "v1"),
+                                   std::make_pair("k2", "v2")));
+  EXPECT_TRUE(m0.ordering_key().empty());
+  EXPECT_TRUE(m0.message_id().empty());
+}
+
+TEST(Message, FromAttributesVectorStdPair) {
+  auto const m0 = Message::FromAttributes({{"k0", "v0"}, {"k1", "v1"}});
+  EXPECT_TRUE(m0.data().empty());
+  EXPECT_THAT(m0.attributes(),
+              UnorderedElementsAre(std::make_pair("k0", "v0"),
+                                   std::make_pair("k1", "v1")));
+  EXPECT_TRUE(m0.ordering_key().empty());
+  EXPECT_TRUE(m0.message_id().empty());
+}
+
+TEST(Message, FromAttributesVectorStdTuple) {
+  using tuple = std::tuple<std::string, std::string>;
+  std::vector<tuple> const attributes({tuple("k1", "v1"), tuple("k2", "v2")});
+  auto const m0 = Message::FromAttributes(attributes);
+  EXPECT_THAT(m0.attributes(),
+              UnorderedElementsAre(std::make_pair("k1", "v1"),
+                                   std::make_pair("k2", "v2")));
+}
+
+TEST(Message, SetData) {
+  auto const m0 = Message::SetData(Message::FromData("original"), "changed");
+  EXPECT_EQ("changed", m0.data());
+}
+
+TEST(Message, SetAttributesIterator) {
+  std::map<std::string, std::string> const attributes(
+      {{"k1", "v1"}, {"k2", "v2"}});
+  auto const m0 = Message::SetAttributes(
+      Message::SetData(Message::FromAttributes({{"k0", "v0"}}), "original"),
+      attributes.begin(), attributes.end());
+  EXPECT_EQ("original", m0.data());
+  EXPECT_THAT(m0.attributes(),
+              UnorderedElementsAre(std::make_pair("k1", "v1"),
+                                   std::make_pair("k2", "v2")));
+}
+
+TEST(Message, SetAttributesVectorStdPair) {
+  std::vector<std::pair<std::string, std::string>> const attributes(
+      {{"k1", "v1"}, {"k2", "v2"}});
+  auto const m0 = Message::SetAttributes(
```

Comment:
Did you consider a fluent builder rather than the static mutator factories? I could imagine this looking like

```cc
auto const m0 = pubsub::MessageBuilder().set_data("original").set_attributes({{"k0, "v0"}}).Build();
```

WDYT?

Reply:
Done.